### PR TITLE
filters out camelCased foreign keys/allows for whitelisting attrs (fixes #56, #70)

### LINKED ISSUE #56: Mapper removes attributes that end in `type` and `id`
I understand this is intended behaviour, but shouldn't there be an option to pass attributes to **not** be omitted from the output?

Another property in the `mapperOptions` param like `{ keepAttributes: ['keep_type', 'thing-id'] }`

### PR DESCRIPTION
**Pull request by jamesdixon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -84,6 +84,7 @@ mapper#map(data, type, mapperOptions)
     - _(required)_ `offset` _(integer)_
     - _(required)_ `limit` _(integer)_
     - _(optional)_ `total` or `rowCount` _(integer)_
+  - _(optional)_ `attrWhitelist` _(string[])_: in accordance with the JSON API spec, the mapper automatically filters out any foreign keys from the primary payload. We currently match on `id`, `-id`, `_id`, `modelId` (camelCase), `-type`, `_type` and `Type` (camelCase). If you need to have an attribute that matches one of the aforementioned patterns pass through, you can add it to the whitelist.
```
Comment (chamini2) [inline README.md]: the documentation says `attrWhitelist` and the code says `attrsWhitelist` (note the **`s`**)

**Comment by chamini2** [comment]
I understand the idea of the regexes to remove attributes, but it never felt great to omit attributes that match certain patterns. 

What do you think of removing this feature and instead offer an opposite option to omit attributes? So everyone has total control of what goes out and what doesn't. Right now the mapper makes a decision based for the user and the user would have to overwrite the functionality of the mapper through the `attrsWhitelist`. 

I prefer the users to pass a list (or even maybe regexes to match like the `restricted` list!) of attributes to omit, that way making obvious the omission. Could save some time debugging.

**Comment by chamini2** [comment]
Also it leaves questions like _why are `Id` and `_id` attrs removed, but `ID` attrs are not?_. And also, these changes could introduce unexpected changes to users that update.

File: README.md (hunk 1 above)
Comment (jamesdixon) [inline README.md]: Good catch

**Comment by jamesdixon** [comment]
@chamini2 I think the `id` conventions are pretty typical, but I agree that it's opinionated. I'd be up for simply adding a restricted list that could be either a RegEx or a string.

**Comment by jamesdixon** [comment]
Also, it's strange: the test that fails on Travis runs just fine for me locally.

**Comment by chamini2** [comment]
> Also, it's strange: the test that fails on Travis runs just fine for me locally.

Apparently bookshelf `0.10.1` breaks this test. I don't know why tho.

**Comment by chamini2** [comment]
> @chamini2 I think the id conventions are pretty typical, but I agree that it's opinionated. I'd be up for simply adding a restricted list that could be either a RegEx or a string.

I'm not sure I understand what you meant, could you explain again please?

**Comment by jamesdixon** [comment]
>> @chamini2 I think the id conventions are pretty typical, but I agree that it's opinionated. I'd be up for simply adding a restricted list that could be either a RegEx or a string.

>I'm not sure I understand what you meant, could you explain again please?

Sorry, I was agreeing with you. Essentially, I was saying that if you'd like to remove the default attr restrictions, that would be fine with me. We should then have some sort of "blacklist" that will omit specific fields. That list should allow for either regular expressions or strings.

**Comment by chamini2** [comment]
Awesome! Yeah, we would need a `omitAttrs` or `attrsBlacklist` option. Could you implement this in another PR?

**Comment by chamini2** [comment]
Let me know if you'll do it, if not, I can.

**Comment by jamesdixon** [comment]
Not sure when I'll get to it -- if you have time and want to implement, be my guest 👍

**Comment by jamesdixon** [comment]
@chamini2 if you do start working on this, just let me know so we don't overlap 👍  I'll do the same

**Comment by chamini2** [comment]
I'll get it done. Closing this one.

**Pull request closed without merging**